### ISSUE TITLE
TODO: Criar um editor de mapas

### ISSUE THREAD

**Issue by lorencinibrdev** [AdoriaStudios/Bomberbro#2]
Caso formos realmente mandar o jogo pro greenlight, seria um baita diferencial empacotar junto um editor de mapas.

**Comment by lorencinibrdev** [comment]
Devido ao estado do projeto, achoq ue é hora de discutirmos esse issue. O que será necessário em um editor de mapas?

Pensei em algo bem simples, como no anexo:

![mapeditor](https://cloud.githubusercontent.com/assets/4778177/6554442/98c9c64a-c63a-11e4-8291-fea068c698ff.png)

Na parte da esquerda, temos a galeria de tiles, que será criada quando, no editor, o usuário abrir uma imagem de 512x512, que será dividida pela engine em tiles de 32x32, e exibidos. Quando o usuário clicar em um tile, qualquer lugar que ele clicar será aplicado esse tile.

Internamente, o editor de mapas será simples: Uma array fixa de 13x11 (posteriormente, podemos fazer uma array dinâmica, e deixar o usuário criar o mapa do tamanho que desejar). Quando for ser salvo, o editor simplesmente escreve um arquivo '.map', seguindo os padrões descritos no Issue https://github.com/AdoriaStudios/Bomberbro/issues/1.

Qualquer sugestão, deve ser inserida aqui.

**Comment by cadumancini** [comment]
Creio que é bem isso mesmo. Nada muito complexo, e suficiente pra ele criar
um mapa legal. No arquivo .map, acredito que daria pra gente colocar uma
linha com o nome do mapa, pra ser exibido na hora da seleção, ou algo do
tipo.

Also, algo que creio que precise ter, é mostrar/fazer o usuário definir os
tipos do tile (parede quebrável, inquebrável, chão, etc.), pra isso entrar
na definição e carregamento dos tiles depois.

Em 9 de março de 2015 09:06, Anderson Lorencini <notifications@github.com>
escreveu:

**Comment by lorencinibrdev** [comment]
Bem pensado. Vamos ter que bolar um jeito de fazer isso dentro da interface do editor de mapas... Até porque essa propriedade do tile é ligada ao tileset...

Quanto as informações do mapa, como nome, descrição e até uma thumbnail, concordo que preciso disso. Vou precisar editar a classe Map do projeto, mas deve ser coisa simples.

**Comment by cadumancini** [comment]
Certo. Thumbnail é bacana também. Daria pra montar ela em tempo de
execução, não? Ou é melhor fazer isso quando o usuário salvar o mapa?

Em 9 de março de 2015 09:26, Anderson Lorencini <notifications@github.com>
escreveu:

**Comment by lorencinibrdev** [comment]
Se não me engano, a SFML tem uma função que cria um .png da tela. Podemos chamar ela assim que o cara salvar o mapa. Aí salvamos em uma pasta dentro do Maps/ do projeto. Aí já era.

A estrutura básica do mapa então vai ser: Um arquivo "MapInfo.map", um arquivo "Thumb.png", e um arquivo "Tileset.png".

**Comment by lorencinibrdev** [comment]
Só confirmando, a SFML tem a função sim. Segue link:

http://www.sfml-dev.org/documentation/2.0/classsf_1_1Image.php#a35d840e67a4303e92e1449c94833faf6

**Comment by cadumancini** [comment]
Certo. Sobre a estrutura do mapa, concordo.

2015-03-09 9:49 GMT-03:00 Anderson Lorencini <notifications@github.com>:

**Comment by lorencinibrdev** [comment]
Code Update:

The basis for the Map Editor is now done.

-> We have a Game State (Accessible from the Main Menu, by pressing the 'M' key).
->The following functionality is already done on the Editor:

'''
* It is capable of loading a tileset from the Folder Structure of the game (Sadly, this functionality does not work on Linux yet). #12 
* The tileset can be used to draw on the map grid, by using the mouse.
'''

The next logical steps are to enable support for Saving and Loading the map, and using the proper map on the gameplay.
